File: 104.maximum-depth-of-binary-tree.cpp
// @leet start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  int maxDepth(TreeNode *root) {
    return depth(root, 1);
  }
	int depth(TreeNode * root, int answer) {
		if (!root) {
			return answer - 1;
		}
		answer++;
		return max(depth(root->left, answer), depth(root->right, answer));
	}
};
// @leet end

File: 104.maximum-depth-of-binary-tree.js
// @leet start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
	return dfs(root, 0);
};
var dfs = function(root, answer) {
	if (!root) {
		return answer--;
	}
	answer++;
	return Math.max(dfs(root.left, answer), dfs(root.right, answer));
};
// @leet end
